Scope is optional with authorization code grant

FStefanni opened the pull request:
## Summary

This pr makes the scope optional with the authorization code grant.

Please note that this code differs from the original pr.
This is due to my understanding of the specification (which could be wrong, so please validate).
The cited RFC section states that scope is *optional*, so, for what I understand:
* If scope is missing: OK, no problem
* If scope is given, then it *must* be valid

Conversely, the original pr did not check the scope at all, so, if a invalid scope where given, it would have passed successfully (which seems to me incorrect).

## Linked issue(s)

Related to #89, point 19, original pr https://github.com/oauthjs/node-oauth2-server/pull/647

@jankapunkt: Other related issues: #84 #79 #71 #64 

## Added tests?

Yes

## OAuth2 standard

RFC 6749 section-4.1.1

jankapunkt commented on lib/grant-types/authorization-code-grant-type.js:
```diff
@@ -176,7 +176,7 @@ AuthorizationCodeGrantType.prototype.revokeAuthorizationCode = function(code) {
   const fns = [
-    this.validateScope(user, client, scope),
+    scope ? this.validateScope(user, client, scope) : scope,
     this.generateAccessToken(client, user, scope),
```
These would still be generated if scope is invalid

jorenvandeweyer replied:
It should be moved outside the promise.all to resolve #84

Uzlopak replied:
So if scope is not null, undefined or an empty string it will be validated, but if it is null or undefined or empty string it will be passed on. 

I also think it should be outside the array, as Promise.all does not have a revert functionality, if validateScope is throwing or returning undefined/empty string. Rename scope from line 177 to parmScope/requestedScope and see that we would not pass the scope from the fns-Array to the other functions but take the unvalidated scope from the parameters.

jankapunkt reviewed (changes requested):
@FStefanni this is correct, scope is optional but if it's invalid there should be no other codes be generated, see #64 #84 so please add these two into consideration.

FStefanni commented:
Hi,

@jankapunkt 
@jorenvandeweyer
@Uzlopak 

I have updated the pr, so please check whether I have understood the comments, or something is still missing.
Btw, I have also found and fixed a bug in `AbstractGrantType.prototype.validateScope()`: it did not return a promise in all cases (but now it does).

Regards

Uzlopak commented:
To be honest: I dont like it.

I expected a minimal change like this:

```javascript
AuthorizationCodeGrantType.prototype.saveToken = function(user, client, authorizationCode, requestedScope) {

  const scope = this.validateScope(user, client, requestedScope);
  const fns = [
      this.generateAccessToken(client, user, validatedScope),
      this.generateRefreshToken(client, user, validatedScope),
      this.getAccessTokenExpiresAt(),
      this.getRefreshTokenExpiresAt()
    ];

  return Promise.all(fns)
    .bind(this)
    .spread(function(accessToken, refreshToken, accessTokenExpiresAt, refreshTokenExpiresAt) {
      const token = {
        accessToken: accessToken,
        authorizationCode: authorizationCode,
        accessTokenExpiresAt: accessTokenExpiresAt,
        refreshToken: refreshToken,
        refreshTokenExpiresAt: refreshTokenExpiresAt,
        scope: scope
      };

      return promisify(this.model.saveToken, 3).call(this.model, token, client, user);
    });
};
```

jorenvandeweyer commented:
I also think you can just use `await` like @Uzlopak suggests. @jwerre is already rewriting the code base to use `await` everywhere else.

jankapunkt commented:
I would wait until it's done with the rewrite and currently focus on integrity for the current state of code. I wil review later today

FStefanni commented:
Hi,

I have re-written by using await, but now 3 tests fail.
I am not able to fix them... can someone help please?

Regards

Uzlopak commented:
I like your change. Good work!

 I would suggest to delete the failing tests, but this could be an inconsistency throughout the codebase. We should wait till everywhere is async await and then merge your change.

jankapunkt commented:
@FStefanni no worries we will check out what's going on here

jankapunkt commented:
@Uzlopak @HappyZombies @jorenvandeweyer should this also be added to `4.2.0`? I think it's not really "breaking" so I would add it to the release.

jorenvandeweyer commented on lib/grant-types/abstract-grant-type.js:
```diff
@@ -97,7 +97,7 @@ AbstractGrantType.prototype.validateScope = async function(user, client, scope)
 
     return validatedScope;
   } else {
-    return scope;
```
not necessary anymore since the function is async now.
